### PR TITLE
Fix for test script EditMenu_Default_UndoRedo

### PR DESCRIPTION
**Pull request by michleza**
*Refactored and fixed the broken test

- moved the test to hydra framework
- added multiple async waits where the ui needed to render before performing test actions
- converted some test logic to a local function for readability
- utilized new script utils files for readability and future qt model

*Re-enabled the test

Signed-off-by: Michael <michleza@amazon.com>

### CODE REVIEW

File: AutomatedTesting/Gem/PythonTests/scripting/EditMenu_Default_UndoRedo.py [hunk 1]
```diff
@@ -44,66 +52,68 @@ def EditMenu_Default_UndoRedo():
...
+        scripting_tools.initialize_sc_editor_objects(self)
+
+        # 3) Open Variable Manager if not opened already
+        scripting_tools.initialize_variable_manager_object(self)
+
+        # 4) Create Graph
+        scripting_tools.create_new_sc_graph(self.sc_editor_main_window)
+
+        # 5) Create new variable
+        add_new_variable_button = self.variable_manager.findChild(QtWidgets.QPushButton, "addButton")
```
Comment (FuzzyCarterAWS) [inline AutomatedTesting/Gem/PythonTests/scripting/EditMenu_Default_UndoRedo.py:86]: We have a const for this button in scripting_constants file.

File: AutomatedTesting/Gem/PythonTests/scripting/EditMenu_Default_UndoRedo.py [hunk 2]
```diff
@@ -44,66 +52,68 @@ def EditMenu_Default_UndoRedo():
...
+        table_view = self.variable_manager.findChild(QtWidgets.QTableView, VARIABLE_PALETTE_QT)
+        model_index = pyside_utils.find_child_by_pattern(table_view, VARIABLE_TYPES[0])
+        # Click on it to create variable
+        pyside_utils.item_view_index_mouse_click(table_view, model_index)
+
+        # 6) Verify if the variable is created initially
+        graph_variables = self.variable_manager.findChild(QtWidgets.QTableView, GRAPH_VARIABLES_QT)
+        result = helper.wait_for_condition(
+            lambda: self.get_graph_variables_row_count(graph_variables) == 1, WAIT_TIME_3) # since we added 1 variable, rowcount=1
+        Report.result(Tests.variable_created, result)
```
Comment (FuzzyCarterAWS) [inline AutomatedTesting/Gem/PythonTests/scripting/EditMenu_Default_UndoRedo.py:98]: Move the `helper.wait_for_condition(...)` into the `Report.result(...)`

File: AutomatedTesting/Gem/PythonTests/scripting/EditMenu_Default_UndoRedo.py [hunk 3]
```diff
@@ -44,66 +52,68 @@ def EditMenu_Default_UndoRedo():
...
+        result = helper.wait_for_condition(
+            lambda: self.get_graph_variables_row_count(graph_variables) == 0, WAIT_TIME_3) # since we triggered undo, rowcount=0
+        Report.result(Tests.undo_worked, result)
+
+        # 8) Trigger Redo action and verify if variable is readded in Variable Manager
+        undo_redo_action = self.sc_editor.findChild(QtWidgets.QAction, "action_Redo")
+        undo_redo_action.trigger()
+        result = helper.wait_for_condition(
+            lambda: self.get_graph_variables_row_count(graph_variables) == 1, WAIT_TIME_3)  # since action is redone 1 variable is readded, rowcount=1
+        Report.result(Tests.redo_worked, result)
```
Comment (FuzzyCarterAWS) [inline AutomatedTesting/Gem/PythonTests/scripting/EditMenu_Default_UndoRedo.py:112]: Move the `helper.wait_for_condition(...)` into the `Report.result(...)`

**Review by FuzzyCarterAWS — CHANGES_REQUESTED** [review]

File: AutomatedTesting/Gem/PythonTests/scripting/EditMenu_Default_UndoRedo.py [hunk 4]
```diff
@@ -44,66 +52,68 @@ def EditMenu_Default_UndoRedo():
...
+        scripting_tools.create_new_sc_graph(self.sc_editor_main_window)
+
+        # 5) Create new variable
+        add_new_variable_button = self.variable_manager.findChild(QtWidgets.QPushButton, "addButton")
+        add_new_variable_button.click()  # Click on Create Variable button
+        # Select variable type
+        table_view = self.variable_manager.findChild(QtWidgets.QTableView, VARIABLE_PALETTE_QT)
+        model_index = pyside_utils.find_child_by_pattern(table_view, VARIABLE_TYPES[0])
+        # Click on it to create variable
+        pyside_utils.item_view_index_mouse_click(table_view, model_index)
```
[the remark this answers is not in the document]
Comment (michleza) [reply AutomatedTesting/Gem/PythonTests/scripting/EditMenu_Default_UndoRedo.py:92]: moved this over to the tools file. Can create a SC variable now by calling the function and providing a variable type.

File: AutomatedTesting/Gem/PythonTests/scripting/EditMenu_Default_UndoRedo.py (hunk 3 above)
Comment (michleza) [reply AutomatedTesting/Gem/PythonTests/scripting/EditMenu_Default_UndoRedo.py]: moved the helper function call into the report.result call

File: AutomatedTesting/Gem/PythonTests/scripting/EditMenu_Default_UndoRedo.py (hunk 1 above)
Comment (michleza) [reply AutomatedTesting/Gem/PythonTests/scripting/EditMenu_Default_UndoRedo.py]: switched this over to the constant

**Review by FuzzyCarterAWS — CHANGES_REQUESTED** [review]

File: AutomatedTesting/Gem/PythonTests/scripting/EditMenu_Default_UndoRedo.py [hunk 5]
```diff
@@ -44,66 +51,62 @@ def EditMenu_Default_UndoRedo():
...
-    variable_manager = sc.findChild(QtWidgets.QDockWidget, "VariableManager")
+        # 1) Open Script Canvas window
+        general.open_pane(SCRIPT_CANVAS_UI)
+        helper.wait_for_condition(lambda: general.is_pane_visible(SCRIPT_CANVAS_UI), WAIT_TIME_3)
 
-    # 4) Create Graph
-    action = pyside_utils.find_child_by_pattern(sc, {"objectName": "action_New_Script", "type": QtWidgets.QAction})
-    action.trigger()
+        # 2) Get the SC window object
+        scripting_tools.initialize_sc_editor_objects(self)
```
[the remark this answers is not in the document]
Comment (michleza) [reply AutomatedTesting/Gem/PythonTests/scripting/EditMenu_Default_UndoRedo.py:76]: moved this over to step 1

File: AutomatedTesting/Gem/PythonTests/scripting/EditMenu_Default_UndoRedo.py [hunk 6]
```diff
@@ -44,66 +51,62 @@ def EditMenu_Default_UndoRedo():
...
-    action.trigger()
-    result = (
-        graph_vars.model().rowCount(QtCore.QModelIndex()) == 1
-    )  # since action is redone 1 variable is readded, rowcount=1
-    Report.result(Tests.redo_worked, result)
+        # 6) Verify if the variable is created initially
+        graph_variables = self.variable_manager.findChild(QtWidgets.QTableView, GRAPH_VARIABLES_QT)
+        # since we added 1 variable, rowcount=1
+        Report.result(Tests.variable_created, helper.wait_for_condition(
+            lambda: self.get_graph_variables_row_count(graph_variables) == 1, WAIT_TIME_3))
```
[the remark this answers is not in the document]
Comment (michleza) [reply AutomatedTesting/Gem/PythonTests/scripting/EditMenu_Default_UndoRedo.py:91]: moved this up to step 5 and reworded the step comment

File: AutomatedTesting/Gem/PythonTests/scripting/EditMenu_Default_UndoRedo.py [hunk 7]
```diff
@@ -44,66 +51,62 @@ def EditMenu_Default_UndoRedo():
...
+        # since we triggered undo, rowcount=0
+        Report.result(Tests.undo_worked, helper.wait_for_condition(
+            lambda: self.get_graph_variables_row_count(graph_variables) == 0, WAIT_TIME_3))
 
+        # 8) Trigger Redo action and verify if variable is re-added in Variable Manager
+        undo_redo_action = self.sc_editor.findChild(QtWidgets.QAction, "action_Redo")
+        undo_redo_action.trigger()
+        # since action is redone 1 variable is re-added, rowcount=1
+        Report.result(Tests.redo_worked, helper.wait_for_condition(
+            lambda: self.get_graph_variables_row_count(graph_variables) == 1, WAIT_TIME_3))
```
[the remark this answers is not in the document]
Comment (michleza) [reply AutomatedTesting/Gem/PythonTests/scripting/EditMenu_Default_UndoRedo.py:105]: turned the integer into a local variable

File: AutomatedTesting/Gem/PythonTests/scripting/scripting_utils/scripting_tools.py [hunk 8]
```diff
@@ -184,6 +185,25 @@ def create_new_sc_graph(sc_editor_main_window):
     create_new_graph_action.trigger()
 
 
+def create_new_variable(self, variable_type):
```
[the remark this answers is not in the document]
Comment (michleza) [reply AutomatedTesting/Gem/PythonTests/scripting/scripting_utils/scripting_tools.py:188]: added error checking to the create variable function

File: AutomatedTesting/Gem/PythonTests/scripting/EditMenu_Default_UndoRedo.py [hunk 9]
```diff
@@ -68,43 +66,42 @@ async def run_test(self):
...
+        # 3) Create Graph
         scripting_tools.create_new_sc_graph(self.sc_editor_main_window)
 
-        # 5) Create new variable
+        # 4) Create and verify the new variable exists in variable manager
         scripting_tools.create_new_variable(self, VARIABLE_TYPES[0])
-
-        # 6) Verify if the variable is created initially
         graph_variables = self.variable_manager.findChild(QtWidgets.QTableView, GRAPH_VARIABLES_QT)
         # since we added 1 variable, rowcount=1
```
Comment (FuzzyCarterAWS) [inline AutomatedTesting/Gem/PythonTests/scripting/EditMenu_Default_UndoRedo.py:83]: Don't really need this comment

File: AutomatedTesting/Gem/PythonTests/scripting/EditMenu_Default_UndoRedo.py [hunk 10]
```diff
@@ -68,43 +66,42 @@ async def run_test(self):
...
+        row_count = 1
         Report.result(Tests.variable_created, helper.wait_for_condition(
-            lambda: self.get_graph_variables_row_count(graph_variables) == 1, WAIT_TIME_3))
+            lambda: self.get_graph_variables_row_count(graph_variables) == row_count, WAIT_TIME_3))
 
-        # 7) Trigger Undo action and verify if variable is removed in Variable Manager
+        # 5) Trigger Undo action and verify if variable is removed in Variable Manager
         undo_redo_action = self.sc_editor.findChild(QtWidgets.QAction, "action_Undo")
         undo_redo_action.trigger()
         # since we triggered undo, rowcount=0
```
Comment (FuzzyCarterAWS) [inline AutomatedTesting/Gem/PythonTests/scripting/EditMenu_Default_UndoRedo.py:91]: Or this one :)

File: AutomatedTesting/Gem/PythonTests/scripting/EditMenu_Default_UndoRedo.py [hunk 11]
```diff
@@ -68,43 +66,42 @@ async def run_test(self):
...
+        row_count = 0
         Report.result(Tests.undo_worked, helper.wait_for_condition(
-            lambda: self.get_graph_variables_row_count(graph_variables) == 0, WAIT_TIME_3))
+            lambda: self.get_graph_variables_row_count(graph_variables) == row_count, WAIT_TIME_3))
 
-        # 8) Trigger Redo action and verify if variable is re-added in Variable Manager
+        # 6) Trigger Redo action and verify if variable is re-added in Variable Manager
         undo_redo_action = self.sc_editor.findChild(QtWidgets.QAction, "action_Redo")
         undo_redo_action.trigger()
         # since action is redone 1 variable is re-added, rowcount=1
```
Comment (FuzzyCarterAWS) [inline AutomatedTesting/Gem/PythonTests/scripting/EditMenu_Default_UndoRedo.py:99]: Or this one.

File: AutomatedTesting/Gem/PythonTests/scripting/scripting_utils/scripting_tools.py [hunk 12]
```diff
@@ -185,21 +186,33 @@ def create_new_sc_graph(sc_editor_main_window):
...
     """
     function for creating a new SC variable through variable manager
 
     param self: the script objecting calling this function
     param variable_type: The variable data type to create as a string. i.e "Boolean"
     returns: none
     """
+
+    if type(new_variable_type) is not str:
+        return Report.critical_result(["Invalid variable type provided", ""], False)
```
Comment (FuzzyCarterAWS) [inline AutomatedTesting/Gem/PythonTests/scripting/scripting_utils/scripting_tools.py:199]: return un-needed

**Review by FuzzyCarterAWS — APPROVED** [review]
Approved after minor changes are finished.

File: AutomatedTesting/Gem/PythonTests/scripting/EditMenu_Default_UndoRedo.py (hunk 9 above)
Comment (FuzzyCarterAWS) [reply AutomatedTesting/Gem/PythonTests/scripting/EditMenu_Default_UndoRedo.py:83]: To be specific, the rowcount comment.

Comment (michleza) [reply AutomatedTesting/Gem/PythonTests/scripting/EditMenu_Default_UndoRedo.py:83]: removed!

File: AutomatedTesting/Gem/PythonTests/scripting/scripting_utils/scripting_tools.py (hunk 12 above)
Comment (michleza) [reply AutomatedTesting/Gem/PythonTests/scripting/scripting_utils/scripting_tools.py:199]: removed this as well

**Review by lsemp3d — APPROVED** [review]

**Pull request merged**